src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -89,73 +89,88 @@ private List<Endpoint> CreateEndpoints()
                     // - Home/Login
                     foreach (var endpointInfo in ConventionalEndpointInfos)
                     {
-                        if (MatchRouteValue(action, endpointInfo, "Area")
-                            && MatchRouteValue(action, endpointInfo, "Controller")
-                            && MatchRouteValue(action, endpointInfo, "Action"))
+                        // An 'endpointInfo' is applicable if:
+                        // 1. it has a parameter (or default value) for 'required' non-null route value
+                        // 2. it does not have a parameter (or default value) for 'required' null route value
+                        var isApplicable = true;
+                        foreach (var routeKey in action.RouteValues.Keys)
                         {
-                            var newPathSegments = endpointInfo.ParsedPattern.PathSegments.ToList();
+                            if (!MatchRouteValue(action, endpointInfo, routeKey))
+                            {
+                                isApplicable = false;
+                                break;
+                            }
+                        }
+
+                        if (!isApplicable)
+                        {
+                            continue;
+                        }
+
+                        var newPathSegments = endpointInfo.ParsedPattern.PathSegments.ToList();
 
-                            for (var i = 0; i < newPathSegments.Count; i++)
+                        for (var i = 0; i < newPathSegments.Count; i++)
+                        {
+                            // Check if the pattern can be shortened because the remaining parameters are optional
+                            //
+                            // e.g. Matching pattern {controller=Home}/{action=Index}/{id?} against HomeController.Index
+                            // can resolve to the following endpoints:
+                            // - /Home/Index/{id?}
+                            // - /Home
+                            // - /
+                            if (UseDefaultValuePlusRemainingSegementsOptional(i, action, endpointInfo, newPathSegments))
```

Comment:
The diff is wonky because of indent, suggest hiding whitespace changes (Diff settings)